pkg/ruler/compat.go

```diff
@@ -60,9 +66,25 @@ func (a *pusherAppender) AppendExemplar(_ uint64, _ labels.Labels, _ exemplar.Ex
 }
 
 func (a *pusherAppender) Commit() error {
+	a.totalWrites.Inc()
+
 	// Since a.pusher is distributor, client.ReuseSlice will be called in a.pusher.Push.
 	// We shouldn't call client.ReuseSlice here.
 	_, err := a.pusher.Push(user.InjectOrgID(a.ctx, a.userID), cortexpb.ToWriteRequest(a.labels, a.samples, nil, cortexpb.RULE))
+
+	if err != nil {
+		reportError := true
+
+		if resp, ok := httpgrpc.HTTPResponseFromError(err); ok && resp.Code/100 == 4 {
+			// Don't report errors that ended with 4xx HTTP status code (series limits, duplicate samples, out of order, etc.)
+			reportError = false
+		}
+
+		if reportError {
+			a.failedWrites.Inc()
+		}
```

Comment:
**suggestion:** If I'm not mistaken, this should be equivalent according to de morgans law and a bit tidier: 
```suggestion
		// Don't report errors that ended with 4xx HTTP status code (series limits, duplicate samples, out of order, etc.)
		if resp, ok := httpgrpc.HTTPResponseFromError(err); !ok || resp.Code/100 != 4 {
			a.failedWrites.Inc()
		}
```

@pstibrany WDYT?

Reply:
Looks good, you know your boolean algebra well 😄